storepool: lock individual storepool details instead of all details

iskettaneh opened the pull request:
This commit takes mutex locks for individual storePool details instead of
the whole storePool map. This should reduce the mutex contention
especially at high store count clusters. Previously, every gossip store
update callback used to lock the whole StorePool.DetailsMu. This blocked
other accesses to any storepool.Detail.

Benchmark results:
```
name                                    old time/op    new time/op    delta
StorePoolReadsWithConcurrentUpdates-12    1.79ms ±40%    0.00ms ± 7%   -99.99%  (p=0.000 n=10+10)

name                                    old alloc/op   new alloc/op   delta
StorePoolReadsWithConcurrentUpdates-12    10.8kB ±57%     0.0kB ±33%   -99.99%  (p=0.000 n=10+10)

name                                    old allocs/op  new allocs/op  delta
StorePoolReadsWithConcurrentUpdates-12      44.2 ±22%       0.0       -100.00%  (p=0.000 n=9+10)
```

Note that the benchmark basically tests that updating a specific store detail
doesn't block accessing other store details.

References: #143145

Release note: None

iskettaneh committed:
storepool: replace StoreDetail map with syncutil.Map

This commit replaces StoreDetail map with syncutil.Map. This will help
reduce contention since syncutil.Map allows mutating an object in the
map without holding the write mutex for the whole map (we still
need to acquire a write mutex for the individual object we are mutating).

This commit just replaces the map access semantics to work with
syncutil.Map. However, we will still be taking the mutex for the whole
map when accessing it. Future commits will fix that.

References: #143145

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144303)
<!-- Reviewable:end -->

iskettaneh committed:
storepool: lock individual storepool details instead of all  details

This commit acquires mutex locks for individual storePool details instead
of the whole storePool map. This should reduce the mutex contention
especially at high store count clusters. Previously, every gossip store
update callback used to lock the whole StorePool.DetailsMu. This blocked
other accesses to any storepool.Detail.

Benchmark results:
```
name                                    old time/op    new time/op    delta
StorePoolReadsWithConcurrentUpdates-12    1.79ms ±40%    0.00ms ± 7%   -99.99%  (p=0.000 n=10+10)

name                                    old alloc/op   new alloc/op   delta
StorePoolReadsWithConcurrentUpdates-12    10.8kB ±57%     0.0kB ±33%   -99.99%  (p=0.000 n=10+10)

name                                    old allocs/op  new allocs/op  delta
StorePoolReadsWithConcurrentUpdates-12      44.2 ±22%       0.0       -100.00%  (p=0.000 n=9+10)
```

Note that the benchmark basically tests that updating a specific store detail
doesn't block accessing other store details.

References: #143145

Release note: None

iskettaneh committed:
kvserver: add BenchmarkStorePoolReadsWithConcurrentUpdates

This commit adds the benchmark: StorePoolReadsWithConcurrentUpdates. It
mainly tries to simulate the storepool mutex contention when there are
periodic updates to other store details concurrent with periodic status
checks for some other store. Future commits will prevent locking the
whole storepool maps for reads/updates. So this benchmark will help
verify that the optimization of only locking a single store details
instead of the whole map works.

References: #143145

Release note: None

iskettaneh committed:
storepool: try to use RLock in StoreDetailMu.status() when possible

Before this commit, we used to take the write lock when querying the
StoreDetailMu.status(). The reason of taking the write lock is that the
StoreDetailMu.status() updates the lastUnavailable timestamp. This
commit attempts to take the read lock, and only upgrades it to a write
lock if we end up needing to update the lastUnavailable timestamp.

References: #143145

Release note: None

kvoli reviewed (commented):
> Reviewed 1 of 11 files at r2, 1 of 15 files at r8, 15 of 15 files at r10, 9 of 9 files at r11, 14 of 14 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144303)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @wenyihu6)

___
*[`-- commits` line 35 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-OOxnmS89y6OYEcwzsvD:-OOxnmS89y6OYEcwzsvE:b-8f1dtg):*
nit: takes or acquires
___
*[`-- commits` line 41 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-OOxntvl7LCTKWAqj_gd:-OOxntvl7LCTKWAqj_ge:b-83sjoh):*
Nice!
___
*[`-- commits` line 53 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-OOxnyGTDQUGYL7BsBkq:-OOxnyGTDQUGYL7BsBkr:bc4ds26):*
nit: this sentence is hard to parse, consider rewording the "basically locks on some store details..." part.
___
*[`pkg/kv/kvserver/allocator/storepool/store_pool.go` line 136 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-OOxZWv-9FXVLAV4Oy9_:-OOxZWv-9FXVLAV4Oy9a:b-rn4pxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5e457a91059de3a8b1ea106fffc2eaa4738a29/pkg/kv/kvserver/allocator/storepool/store_pool.go#L136)):*
> ```Go
> 	}
> 
> 	// Note: If Desc needs to be deep copied in the future, add:
> ```

This comment seems stale, this commit is adding in the logic to deep copy the descriptor when non-nil.
___
*[`pkg/kv/kvserver/allocator/storepool/store_pool.go` line 434 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-OOxo_GH7Be4lCzFcVq7:-OOxo_GH7Be4lCzFcVq8:b-t11j4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/20987388352c5a8d838aa3cf48217425f053b917/pkg/kv/kvserver/allocator/storepool/store_pool.go#L434)):*
> ```Go
> 	ids := make(roachpb.StoreIDSlice, sp.getStoreDetailsCount())
> 	sp.Details.StoreDetails.Range(func(id roachpb.StoreID, sd *StoreDetailMu) bool {
> 		sd.RLock()
> ```

Is this `RLock` necessary, aren't we only grabbing the key here?
___
*[`pkg/kv/kvserver/allocator/storepool/store_pool.go` line 784 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-OOxpF531Jiqp1Ed6bMn:-OOxpF531Jiqp1Ed6bMo:b6yo4bf) ([raw file](https://github.com/cockroachdb/cockroach/blob/20987388352c5a8d838aa3cf48217425f053b917/pkg/kv/kvserver/allocator/storepool/store_pool.go#L784)):*
> ```Go
> 		return false, 0, errors.Errorf("store %d was not found", storeID)
> 	}
> 	sd.Lock()
> ```

nit: although it was previously Locking, this could be updated to an RLock.


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144303)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @kvoli, and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/storepool/store_pool.go` line 136 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-OOxZWv-9FXVLAV4Oy9_:-OOyWguyDAJGBIP6Hou1:b-pv5ypx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5e457a91059de3a8b1ea106fffc2eaa4738a29/pkg/kv/kvserver/allocator/storepool/store_pool.go#L136)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

This comment seems stale, this commit is adding in the logic to deep copy the descriptor when non-nil.
</blockquote></details>

Yeah that was a comment for myself that I forgot to remove.
___
*[`pkg/kv/kvserver/allocator/storepool/store_pool.go` line 434 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-OOxo_GH7Be4lCzFcVq7:-OOyVZSz6vRUtxVpckDd:b-aru2gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/20987388352c5a8d838aa3cf48217425f053b917/pkg/kv/kvserver/allocator/storepool/store_pool.go#L434)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Is this `RLock` necessary, aren't we only grabbing the key here?
</blockquote></details>

Good catch! Yeah I am not sure why I was grabbing the RLock.
Done.
___
*[`pkg/kv/kvserver/allocator/storepool/store_pool.go` line 784 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-OOxpF531Jiqp1Ed6bMn:-OOyX_7T86tzzEQfwwK5:b-yxu0un) ([raw file](https://github.com/cockroachdb/cockroach/blob/20987388352c5a8d838aa3cf48217425f053b917/pkg/kv/kvserver/allocator/storepool/store_pool.go#L784)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: although it was previously Locking, this could be updated to an RLock.
</blockquote></details>

Agreed. Done


<!-- Sent from Reviewable.io -->

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 9 files at r15, 13 of 14 files at r16, 1 of 1 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144303#-:-OP0mhq02Rgo4lx1p0Q6:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @wenyihu6)



<!-- Sent from Reviewable.io -->

rafiss commented:
This mutex showed up in the [mutex profile](https://github.com/user-attachments/files/19170317/mutex.pb.gz) that I grabbed by running `./dev bench ./pkg/sql/tests --filter BenchmarkParallelSysbench --test-args '-test.mutexprofile=mutex.pb.gz  -test.cpu=8'` a few weeks ago.

I'm curious if we see an overall improvement on sysbench with this change.
